processor/src/trace/virtual_bus/sum_check/prover/mod.rs

```diff
@@ -0,0 +1,289 @@
+{
+    composition_poly: P,
+    eval_domain: EvaluationDomain<E>,
+    _challenger: PhantomData<C>,
+}
+
+impl<B, E, P, C, H> SumCheckProver<B, E, P, C, H>
+where
+    B: StarkField,
+    E: FieldElement<BaseField = B>,
+    P: CompositionPolynomial<E>,
+    C: RandomCoin<Hasher = H, BaseField = B>,
+    H: ElementHasher<BaseField = B>,
+{
+    /// Constructs a new [SumCheckProver] given a multivariate composition polynomial.
+    /// The multivariate composition polynomial corresponds to the `g` polynomial in the
+    /// description of the [SumCheckProver] struct.
+    pub fn new(composition_poly: P) -> Self {
+        let max_degree = composition_poly.max_degree();
+        let eval_domain = EvaluationDomain::new(max_degree);
+
+        Self {
+            composition_poly,
+            eval_domain,
+            _challenger: PhantomData,
+        }
+    }
+
+    /// Given an initial claim `claim`, a mutable vector of multi-linear polynomials `mls` and
+    /// a number of rounds `num_rounds`, computes `num_rounds` iterations of the sum-check protocol
+    /// starting from claim `claim`.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - No multi-linears were provided.
+    /// - Number of rounds is zero or is greater than the number of variables of the multilinears.
+    /// - The provided multi-linears have different arities.
+    pub fn prove(
+        &self,
+        claim: E,
+        mls: &mut [MultiLinear<E>],
+        num_rounds: usize,
+        coin: &mut C,
+        //) -> Result<(RoundClaim<E>, Proof<E>), Error> {
+    ) -> Result<Proof<E>, Error> {
+        // there should be at least one multi-linear polynomial provided
+        if mls.is_empty() {
+            return Err(Error::NoMlsProvided);
+        }
+
+        // there should be at least one round to prove
+        if num_rounds == 0 {
+            return Err(Error::NumRoundsZero);
+        }
+
+        // there can not be more rounds than variables of the multi-linears
+        let ml_variables = mls[0].num_variables();
+        if num_rounds > ml_variables {
+            return Err(Error::TooManyRounds);
+        }
+
+        // there should at least be one variable for the protocol to be non-trivial
+        if ml_variables < 2 {
+            return Err(Error::AtLeastOneVariable);
+        }
+
+        // all multi-linears should have the same arity
+        if !mls.iter().all(|ml| ml.num_variables() == ml_variables) {
+            return Err(Error::MlesDifferentArities);
+        }
+
+        let mut round_proofs = vec![];
+
+        // setup first round claim
+        let mut current_round_claim = RoundClaim {
+            eval_point: vec![],
+            claim,
+        };
+
+        // run the first round of the protocol
+        let mut round_proof = sumcheck_round(&self.composition_poly, mls);
+        round_proofs.push(round_proof);
+        // reseed with the s_0 polynomial
+        coin.reseed(H::hash_elements(&round_proofs[0].poly_evals));
```

Comment:
nit: I would first reseed the coin using `&round_proof.poly_evals`, and then push `round_proof` in `round_proofs`

Reply:
Makes sense